### PR TITLE
docs: Doxygen

### LINKED ISSUE #29: Epic: improve documentation
The documentation needs to be improved.

- [x] publish it to an own dominion arxlang.org
- [ ] #44
- [ ] #43
- [x] #42
- [ ] #41
- [ ] #48

ref: https://arxlang.org

### PR DESCRIPTION
**Pull request by luabida**
Adding docstring to classes and methods.
fix #29 partially

# PR Checklist

Ensure that:

- [ ] it includes tests.
- [ ] the tests for your implementation are executed on CI.
- [x] it includes documentation
- [x] the PR's title is according to the [semantic-release pattern][semantic-release-message].
  - `build`: Changes that affect the build system or external dependencies (example scopes: cmake, meson, etc)
  - `ci`: Changes to our CI configuration files and scripts (examples: CircleCi, SauceLabs)
  - `docs`: Documentation only changes
  - `perf`: A code change that improves performance
  - `refactor`: A code change that neither fixes a bug nor adds a feature
  - `test`: Adding missing tests or correcting existing tests
  - `chore:` Can be used as a generic task for tasks such as CI, test, support tasks or any other task that is not user facing task.
  - `feat:` is used for a new feature or when a existent one is improved.
  - `fix:` is used when a bug is fixed.
  - `breaking change:` is used when there is a compability break.
- [x] pre-commit hooks were executed locally

[semantic-release-message]: https://github.com/semantic-release/semantic-release/blob/master/README.md#commit-message-format "Semantic Release Commit Message Format"

### CODE REVIEW

File: arx/src/parser.cpp [hunk 1]
```diff
@@ -167,7 +195,7 @@ std::unique_ptr<ForExprAST> ParseForExpr() {
     return nullptr;
   }
 
-  // The step value is optional.
+  /** The step value is optional. */
```
Comment (xmnlab) [inline arx/src/parser.cpp]: I think that for the regular comments it is ok to keep it with `//`
any reason for changing that?

PS: there are a lot of other situations like this. I am just commenting in this one

File: docs/templates/example_test.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
+/**
```
Comment (xmnlab) [inline docs/templates/example_test.cpp:1]: do we really need this example file?

File: arx/src/codegen.cpp [hunk 3]
```diff
@@ -1,12 +1,8 @@
-// note: arrow will not be used yet
-// #include <arrow/api.h>
-// #include <arrow/csv/api.h>
-// #include <arrow/io/api.h>
-// #include <arrow/ipc/api.h>
-// #include <arrow/pretty_print.h>
-// #include <arrow/result.h>
-// #include <arrow/status.h>
-// #include <arrow/table.h>
+*note : arrow will not be used yet * #include<arrow / api.h> *
```
Comment (xmnlab) [inline arx/src/codegen.cpp]: can you move back to the original structure please?

File: arx/src/codegen.cpp [hunk 4]
```diff
@@ -60,16 +56,17 @@
 #include "settings.h"
 #include "utils.h"
 
-extern std::map<char, int> BinopPrecedence;
+    extern std::map<char, int> BinopPrecedence;
```
Comment (xmnlab) [inline arx/src/codegen.cpp]: could you remove this extra space?

File: arx/src/codegen.cpp [hunk 5]
```diff
@@ -81,19 +78,28 @@ std::map<std::string, llvm::AllocaInst*> NamedValues;
 std::unique_ptr<llvm::orc::ArxJIT> TheJIT;
 std::map<std::string, std::unique_ptr<PrototypeAST>> FunctionProtos;
 
-//===----------------------------------------------------------------------===//
-// Debug Info Support
-//===----------------------------------------------------------------------===//
-
+/**
```
Comment (xmnlab) [inline arx/src/codegen.cpp]: can you keep regular comments with `//` please?

**Review by xmnlab — CHANGES_REQUESTED** [review]
thanks for working on that @luabida !

just a few comments:

- please keep regular comments with `//`
- check if no cpp/h files includes were not changed
- you can remove the test file for the documentation that I mentioned in a comment inline
- check if no regular code was changed (as the includes I mentioned above)


how can I test it locally?

I think that a normal workflow for testing would be:
- run make docs-build
- in another terminal, run make docs-api

File: arx/src/parser.cpp (hunk 1 above)
Comment (luabida) [reply arx/src/parser.cpp]: Thanks for the review! Ok, I'll change it back

**Comment by luabida** [comment]
I've changed back the comments mentioned and solved some warnings from Doxygen.
To test the API Docs, first urn `make docs-api` and access with a browser the html page located at `{...}/arx/build/api/index.html`

**Comment by xmnlab** [comment]
https://github.com/arxlang/arx/actions/runs/3143836541/jobs/5109060932#step:8:1

@luabida the lint step is failing .. do you mind fix that ?
you can use `pre-commit run --all-files --verbose`

**Comment by luabida** [comment]
> @luabida the lint step is failing .. do you mind fix that ? you can use `pre-commit run --all-files --verbose`

sorry for that, I've run lint but forgot to push it. Yet, the lint is still failing to me, maybe you can catch the error.

**Review by xmnlab — APPROVED** [review]
LGTM! thanks for working on that!

**Pull request merged**